File: app/src/main/java/net/hogelab/android/vpndns/domain/repository/DnsRepository.kt
package net.hogelab.android.vpndns.domain.repository

import kotlinx.coroutines.flow.StateFlow
import net.hogelab.android.vpndns.domain.model.BlacklistEntry
import net.hogelab.android.vpndns.domain.model.DnsEntry
import net.hogelab.android.vpndns.domain.model.HistorySortConfig

/**
 * DNS 履歴とブラックリストを統合管理するリポジトリ
 */
interface DnsRepository {
    /**
     * DNS 履歴 (ブラックリストの状態が反映され、ソート設定に従った順序)
     */
    val history: StateFlow<List<DnsEntry>>

    /**
     * ブラックリスト一覧
     */
    val blacklist: StateFlow<List<BlacklistEntry>>

    /**
     * 現在のソート設定
     */
    val sortConfig: StateFlow<HistorySortConfig>

    /**
     * 履歴にホスト名を追加または更新する
     */
    fun addHistory(host: String)

    /**
     * 履歴をクリアする
     */
    fun clearHistory()

    /**
     * ブラックリストに登録する
     */
    fun addToBlacklist(hostName: String)

    /**
     * ブラックリストから削除する
     */
    fun removeFromBlacklist(hostName: String)

    /**
     * 指定されたホストがブロック対象かどうかを判定する
     */
    fun isBlocked(hostName: String): Boolean

    /**
     * ブラックリストをクリアする
     */
    fun clearBlacklist()

    /**
     * ソート設定を変更する
     */
    fun setSortConfig(config: HistorySortConfig)
}

File: mock-ios/VpnDns/domain/repository/DnsRepository.swift
import Foundation

/// DNSリクエストの履歴および統計データを管理するリポジトリのインターフェース
public protocol DnsRepository {
    /// 記録されているすべてのDNS解決履歴のリスト
    var history: [DnsEntity] { get }
    
    /// 新しいDNS解決結果を履歴に追加、または既存の記録をインクリメントする
    func add(_ entity: DnsEntity)
    
    /// すべての解決履歴をクリアする
    func clear()
    
    /// 登録されているすべてのブラックリストルールのリスト
    var blacklist: [BlacklistEntity] { get }
    
    /// ブラックリストに新しいルールを追加する
    func add(_ entity: BlacklistEntity)
    
    /// ブラックリストから特定のルールを削除する
    func remove(_ entity: BlacklistEntity)
    
    /// すべてのブラックリストルールをクリアする
    func clearBlacklist()
}

